source/code/plugins/in_npmd_server.rb

```diff
@@ -0,0 +1,428 @@
+
+            @stop_sync.synchronize do
+                # Reading standard error
+                @writer.close
+                _arr = Array.new
+                while _line = @reader.gets
+                    _h = Hash.new
+                    _h["Message"] = _line
+                    _arr << _h
+                end
+
+                @npmdExitHandled = true
+                @wait_to_collect_stderror.broadcast()
+
+                emit_error_log_dataitems(_arr) unless @npmdIntendedStop
+            end
+
+            # Checking if NPMD exited as planned
+            if @npmdIntendedStop
+                Logger::logInfo "NpmdAgent exited normally"
+            else
+                log_error "NpmdAgent ended with exit status #{_exitRes[1]}"
+                Logger::logInfo "Restarting NPMD"
+                start_npmd_async_retry_thread()
+            end
+        end
+
+        def start_npmd
+            unless is_npmd_seen_in_ps()
+                @reader, @writer = IO.pipe
+                @npmdIntendedStop = false
+                @npmdProcessId = Process.spawn(@binary_invocation_cmd, :err=>@writer)
+                _t = Thread.new(&method(:handle_exit))
+            else
+                Logger::logInfo "Npmd already seen in PS"
+            end
+        end
+
+        def stop_npmd
+            if defined?@npmdProcessId and is_npmd_seen_in_ps()
+                @stop_sync.synchronize do
+                    @npmdIntendedStop = true
+                    Process.kill(STOP_SIGNAL, @npmdProcessId.to_i)
+                end
+            else
+                Logger::logInfo "NPMD agent found already stopped"
+            end
+        end
+
+        def send_config
+            begin # checking for File call errors with this config file
+                unless File.exist?(@location_control_data)
+                    Logger::logWarn "No file #{@location_control_data} found at location"
+                else
+                    if defined?@npmdClientSock and !@npmdClientSock.nil?
+
+                        # Read the UI configuration from file location
+                        _uiXml = File.read(@location_control_data)
+                        if _uiXml.nil? or _uiXml == ""
+                            Logger::logWarn "File read at #{@location_control_data} got nil or empty string"
+                            return
+                        end
+
+                        # Transform the UI XML configuration to agent configuration
+                        _agentConfig = NPMDConfig::GetAgentConfigFromUIConfig(_uiXml)
+                        if _agentConfig.nil? or _agentConfig == ""
+                            Logger::logWarn "Agent configuration transformation returned empty"
+                            return
+                        end
+
+                        @npmdClientSock.puts _agentConfig
+                        @npmdClientSock.flush
+                        @num_config_sent += 1 unless @num_config_sent.nil?
```

Comment:
you need to initialize the num_config_sent. if it was nil, it would never be assigned a value.